### ISSUE TITLE
Capability revocation for PB

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7202]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#114|
|Type|enhancement|
|Created|2003-07-21 19:54:13Z|

<details><summary>Searchable metadata</summary>

```
trac-id__114 114
type__enhancement enhancement
reporter__radix radix
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__pb pb
keywords__ 
time__1058817253000000 1058817253000000
changetime__1465027210167161 1465027210167161
version__ 
owner__ 
cc__radix cc__warner
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
PB needs a way to revoke capabilities on a more fine-grained
scale than simply disconnecting a client. 

e.g., pb.Referenceable (and others) could have a
revokeFrom(client) method, accessible only from the "sending
side".

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
Again, part of New PB.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
hmm. this wouldn't be too hard to do, it would just remove the clid from the
Broker's table. You'd have to make sure that they can't just re-fetch the
reference using whatever they did earlier (i.e. actually using this feature
securely may not be trivial).

Tell you what, write me a test case (see sandbox/warner/test_pb.py for a start)
and I'll implement the feature. And decide whether you want it to be 

 ref.revokeFrom(broker)

or

  broker.revoke(ref)

and also decide if the client should get their notifyOnDisconnect() watcher
triggered, or if there should be a new "you've been revoked" callback, or if
they shouldn't find out at all until they actually send a message.

Oh, and come up with a clever way for a bad guy to work around it, since I'm
sure there'll be one, and we should try to close the hole.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set status to `closed`|
|-|-|

I don't actually think that there's much value to PB doing this (since it would be hard for code to get a reference to an interesting thing to revoke).  I think this is the analogous ticket in the Foolscap tracker: <http://foolscap.lothar.com/trac/ticket/44>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli commented|
|-|-|

In changeset a403e0e1a399dd3df12fad770e76b0c64c117b4f
```
#!CommitTicketReference repository="" revision="a403e0e1a399dd3df12fad770e76b0c64c117b4f"
Merge pull request #114 from rodrigc/8398-rodrigc-print: use print function instead of print statement

Author: rodrigc
Reviewer: cyli
Fixes: #8398

Convert from print statement to print() function in twisted/tap
```
